homeassistant/components/leafspy/device_tracker.py

```diff
@@ -0,0 +1,206 @@
+"""Device tracker platform that adds support for Leaf Spy."""
+import logging
+
+from homeassistant.core import callback
+from homeassistant.const import (
+    ATTR_LATITUDE,
+    ATTR_LONGITUDE,
+    ATTR_BATTERY_LEVEL,
+)
+from homeassistant.components.device_tracker.const import (
+    ENTITY_ID_FORMAT, ATTR_SOURCE_TYPE, SOURCE_TYPE_GPS)
+from homeassistant.components.device_tracker.config_entry import (
+    DeviceTrackerEntity
+)
+from homeassistant.util import slugify
+from homeassistant.helpers.restore_state import RestoreEntity
+from homeassistant.helpers import device_registry
+from .const import DOMAIN as LS_DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+PLUG_STATES = [
+    "Not Plugged In",
+    "Partially Plugged In"
+    "Plugged In"
+]
+
+CHARGE_MODES = [
+    "Not Charging",
+    "Level 1 Charging (100-120 Volts)",
+    "Level 2 Charging (200-240 Volts)",
+    "Level 3 Quick Charging"
+]
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up Leaf Spy based off an entry."""
+    async def _receive_data(dev_id, **data):
+        """Receive set location."""
+        entity = hass.data[LS_DOMAIN]['devices'].get(dev_id)
+
+        if entity is not None:
+            entity.update_data(data)
```

Comment:
It's not safe to call entity state update methods from outside the entity in non polling platforms. State update is only valid after the entity has been added to home assistant. We can't guarantee that here.

Please use our dispatch helper and connect this method to a signal in `async_added_to_hass`. Here instead send the signal.